library/std/src/rt.rs

```diff
@@ -28,14 +28,31 @@ fn lang_start_internal(
     use crate::panic;
     use crate::sys_common;
 
-    // SAFETY: Only called once during runtime initialization.
-    unsafe { sys_common::rt::init(argc, argv) };
+    unsafe {
+        // SAFETY: Only called once during runtime initialization.
+        sys_common::rt::init(argc, argv);
+    }
 
     let exit_code = panic::catch_unwind(main);
-
     sys_common::rt::cleanup();
-
-    exit_code.unwrap_or(101) as isize
+    match exit_code {
+        Ok(exit_code) => exit_code as isize,
+        Err(dangerous_panic_payload) => {
+            // Dropping the `dangerous_panic_payload` will return control back to the user code.
+            // This user code has the freedom to (attempt to) panic again. On the other hand,
+            // unwinding out of this function is UB. This is a requirement imposed by a combination
+            // of how the `#[lang="start"]` attribute is implemented as well as the implementation
+            // of the panicking mechanism.
+            //
+            // Since the only way to reach this branch is by panicking in the first place, we take
+            // a liberty of running the drop in an environment that looks as if a panic is already
+            // in progress. This way any panics in the user code that we will run shortly will
+            // become double-panic aborts.
+            panic_count::increase();
```

Comment:
Maybe it would be safer to add a drop bomb that aborts the process when unwinding above the `catch_unwind(main)` call? That would also ensure an abort when C++ unwinds into rust code. We define that as UB anyway, but just in case.

Reply:
There are multiple implementation approaches that could be taken here, indeed. Each has their own considerations.

I don't think the choice should be made based on behaviour in case UB is invoked, however. This code handling a super niche corner-case in the first place I think the simplicity of the implementation is probably outweighs in benefit the better UB resiliency.

That said, a drop-guard would allow to guard against panics in `rt::init` and `rt::cleanup`, which is quite desirable IME.

Reply:
Alas, the drop guard approach does _not_ work. That's because there isn't any catch pads the unwinding implementation could find, and thus you get the same old UB as if we did nothing at all.

I did adjust the implementation to use another `catch_unwind`, however, which should handle the issues with user code and certain runtime functions panicking.